### ISSUE TITLE
[akka/pekko] Too many paths created during record partitioning

### ISSUE THREAD

**Issue by sndnv** [mjakubowski84/parquet4s#332]
We've been investigating some OoM errors in our services and we found that a lot of Hadoop paths are being created by [this](https://github.com/mjakubowski84/parquet4s/blob/v2.15.0/akkaPekko/src/main/scala/com/github/mjakubowski84/parquet4s/ParquetPartitioningFlow.scala#L277) function:

```scala
 private def partition(record: RowParquetRecord): (Path, RowParquetRecord) =
      partitionBy.foldLeft(basePath -> record) { case ((currentPath, currentRecord), partitionPath) =>
        currentRecord.removed(partitionPath) match {
          case (Some(BinaryValue(binary)), modifiedRecord) =>
            Path(currentPath, s"$partitionPath=${binary.toStringUsingUTF8}") -> modifiedRecord
          
            // ... various failure cases
        }
      }
```

What happens here is that, for each record, the final path is calculated and the partitioning columns are removed from the original record. In our case, we have 5 partitioning fields resulting in 5 [calls](https://github.com/mjakubowski84/parquet4s/blob/v2.15.0/core/src/main/scala/com/github/mjakubowski84/parquet4s/Path.scala#L21) to create new Hadoop paths. It looks like Hadoop underneath does a lot of string manipulation to get these paths constructed and all that results in a bunch of string builders, arrays of chars and java URIs being created.

On its own, this is not too bad but we have instances processing 15K records per second, resulting in 75K paths per second being created this way and it seems the garbage collector is having trouble keeping up. I doubt that this is the source of the OoM errors but it definitely adds to the memory pressure.

Instead, it should be enough to create just one path per record, at the end of the traversal of the partitioning fields (PR to follow).

**Comment by mjakubowski84** [comment]
@sndnv thank you for the contribution! I am away on vacation till February. I will have a look at the PR when I am back.

**Comment by mjakubowski84** [comment]
@sndnv I refactored the logic a little bit more for even better performance.
The changes are released in 2.16.1
